src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,160 @@
-      term_info.detach_term(self);
-    }
-
-    term_info.merge_scope(parent_scope);
 
-    detach
+      _ => false,
+    })
   }
 
-  fn float_app(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::App { fun, arg, .. } = self else { unreachable!() };
-
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    let fun_detach = fun.go_float(depth + 1, term_info);
-    let fun_scope = term_info.replace_scope(BTreeSet::new());
-
-    let arg_detach = arg.go_float(depth + 1, term_info);
-    let arg_scope = term_info.replace_scope(parent_scope);
+  /// Checks if the term is a lambda sequence with the body being a variable in the scope or a reference.
+  fn is_constant_lambda(&self, book: &Book, seen: &mut IndexSet<Name>) -> bool {
+    let mut current = self;
+    let mut scope = Vec::new();
 
-    let detach = match fun_detach {
-      // If the fun scope is not empty, we know the recursive ref is not in a active position,
-      // Because if it was an application, it would be already extracted
-      //
-      // e.g.: {recursive_ref some_var}
-      Detach::Recursive if !fun_scope.is_empty() => arg_detach,
-
-      Detach::Recursive if arg_scope.is_empty() && arg_detach.can_detach() => {
-        term_info.detach_term(self);
-        Detach::Combinator
-      }
-
-      // If the only term that is possible to detach is just a Term::Ref,
-      // there is no benefit to it, so that case is skipped
-      Detach::Recursive if !matches!(fun, box Term::Ref { .. }) => {
-        term_info.detach_term(fun);
-        arg_detach
+    while let Term::Lam { nam, bod, .. } = current {
+      if let Some(nam) = nam {
+        scope.push(nam);
       }
-
-      _ => fun_detach & arg_detach,
-    };
-
-    term_info.merge_scope(fun_scope);
-    term_info.merge_scope(arg_scope);
-
-    detach
-  }
-
-  fn float_mat(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::Mat { args, rules } = self else { unreachable!() };
-
-    let mut detach = Detach::Combinator;
-    for arg in args {
-      detach = detach & arg.go_float(depth + 1, term_info);
+      current = bod;
     }
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    for rule in rules.iter_mut() {
-      for pat in &rule.pats {
-        debug_assert!(pat.is_native_num_match());
-      }
-
-      let arm_detach = match rule.body.go_float(depth + 1, term_info) {
-        // If the recursive ref reached here, it is not in a active position
-        Detach::Recursive => Detach::Combinator,
-        detach => detach,
-      };
 
-      detach = detach & arm_detach;
+    match current {
+      Term::Var { nam } if scope.contains(&nam) => true,
+      Term::Ref { .. } => true,
+      term => term.is_safe(book, seen),
     }
+  }
 
-    term_info.merge_scope(parent_scope);
-    detach
+  /// A term is a combinator if it is a lambda abstraction without free variables.
+  pub fn is_combinator(&self) -> bool {
```

Comment:
A combinator is any term without free variables (+ free unscoped lambda abstractions in case of our modified lambda logic).

You shouldn't be floating only combinators that are abstractions or applications, but any combinator.

e.g.
```
@x (x match (Foo 0) {...})
```
The whole match expression could be extracted since it's a combinator.
I'm not sure how useful that is, but technically a combinator if all its children are combinators.

Reply:
The match expressions are converted to lambdas before this pass, so it will probably be extracted

Reply:
Native matches won't